src/blocks/keyboard_layout/xkb_event.rs

```diff
@@ -0,0 +1,151 @@
+    let colon_i = buf.iter().position(|c| *c == b':').unwrap_or(buf.len());
+    let layout = buf[..colon_i]
+        .split(|&c| c == b'+')
+        .skip(1) // layout names start from index 1
+        .nth(index)
+        .error("Index out of range")?;
+    std::str::from_utf8(layout).error("non utf8 layout")
+}
+
+async fn get_layout(connection: &RustConnection) -> Result<String> {
+    let xkb_state = connection
+        .xkb_get_state(ID::USE_CORE_KBD.into())
+        .await
+        .error("xkb_get_state failed")?
+        .reply()
+        .await
+        .error("xkb_get_state reply failed")?;
+    let group: u8 = xkb_state.group.into();
+
+    let symbols_name = connection
+        .xkb_get_names(
+            ID::USE_CORE_KBD.into(),
+            NameDetail::GROUP_NAMES | NameDetail::SYMBOLS,
+        )
+        .await
+        .error("xkb_get_names failed")?
+        .reply()
+        .await
+        .error("xkb_get_names reply failed")?
+        .value_list
+        .symbols_name
+        .error("symbols_name is empty")?;
+
+    let name = connection
+        .get_atom_name(symbols_name)
+        .await
+        .error("get_atom_name failed")?
+        .reply()
+        .await
+        .error("get_atom_name reply failed")?
+        .name;
+    let layout = parse_layout(&name, group as _)?;
+
+    Ok(layout.to_owned())
+}
+
+async fn prefetch_xkb_extension(connection: &RustConnection) -> Result<UseExtensionReply> {
+    connection
+        .prefetch_extension_information(xkb::X11_EXTENSION_NAME)
+        .await
+        .error("prefetch_extension_information failed")?;
+
+    let reply = connection
+        .xkb_use_extension(XCB_XKB_MAJOR_VERSION, XCB_XKB_MINOR_VERSION)
+        .await
+        .error("xkb_use_extension failed")?
+        .reply()
+        .await
+        .error("xkb_use_extension reply failed")?;
+
+    Ok(reply)
+}
+
+impl XkbEvent {
+    pub(super) async fn new() -> Result<Self> {
+        let (connection, _, drive) = RustConnection::connect(None)
+            .await
+            .error("Failed to open XCB connection")?;
+
+        tokio::spawn(drive);
+        let reply = prefetch_xkb_extension(&connection)
+            .await
+            .error("Failed to prefetch xkb extension")?;
+
+        if !reply.supported {
+            return Err(Error::new(
+                "This program requires the X11 server to support the XKB extension",
+            ));
+        }
+
+        connection
+            .xkb_select_events(
+                ID::USE_CORE_KBD.into(),
+                EventType::default(),
+                EventType::STATE_NOTIFY,
+                MapPart::default(),
+                MapPart::default(),
+                &SelectEventsAux::new(),
+            )
+            .await
+            .error("Failed to select events")?;
+
+        let cur_layout = get_layout(&connection)
+            .await
+            .error("Failed to get current layout")?;
```

Comment:
I think there is no need to store current layout in the struct and this line can be moved down to `get_info` implementation. This would also simplify `wait_for_change`.

Reply:
so you mean moving the whole logic of `wait_for_change` to `get_info`? because calling `get_layout` function has a cost and we only wanna call it for specific events (`xkb::StatePart::GROUP_STATE`)

Reply:
`get_info` is only called when `wait_for_change` returns and on the first render. So it is okay to perform costly operations in it.

Reply:
So instead of calling `get_layout` from two locations we can call it from only one.